Decouple Nomination and Ledger classes

AndrejMitrovic opened the pull request:
Includes some commits from PRs currently in review. I'll remove them as soon as those PRs get merged.

The commits which are not marked `[removed]` may be reviewed.

Requires: ~https://github.com/bpfkorea/agora/pull/576~

Fixes #572

AndrejMitrovic committed:
Move nomination responsibility to Nominator class

The Ledger class suffered from leaky abstractions,
acting as both a ledger-tracking class and as a
nominating class, conflating two different
responsibilities.

The previous several commits have gradually shifted
that responsibility to the Nominator class.

This will also make it easier to change the nomination
protocol to be timer-based, rather than trigger-based
as it is now.

Fixes #572

AndrejMitrovic committed:
Rename retrieveLatestBlocks => startPeriodicCatchup

It more closely reflects the purpose of the function.

AndrejMitrovic committed:
Decouple isNominating() call

The nomination busy state will be moved
to the Nomination class, after which
the isNominating() delegate will query
the Nominator in place of the Ledger.

To make commits easier to review the change is
made in several commits rather than a big one.

AndrejMitrovic committed:
Move SCP quorum conversion function to Nominator

This further reduces the responsibility of the Node class.

AndrejMitrovic committed:
Nominator: Make network setup a separate method

It decouples the Node from having to know how the
Nominator chooses the clients to communicate with.

Also moved the initialization of the Nominator to the
constructor of the Node.

AndrejMitrovic committed:
Ledger: Split off function preparing transactions

The tryNominateTXSet() will be removed once the
Nominator is in charge of starting the nominating
process in place of the Ledger.

AndrejMitrovic commented:
Unblocked!

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -80,6 +81,9 @@ public extern (C++) class Nominator : SCPDriver
     private ulong[TimerType.max + 1] active_timer_ids;
 
+    /// Quorum config
+    private QuorumConfig quorum_conf;
```
```suggestion
    private const QuorumConfig quorum_conf;
```

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -91,17 +95,18 @@ extern(D):
-        TaskManager taskman, NetworkClient[PublicKey] peers,
-        SCPQuorumSet quorum_set)
+    public this (KeyPair key_pair, Ledger ledger, TaskManager taskman,
+        NetworkClient[PublicKey] peers, QuorumConfig config)
```
```suggestion
        NetworkClient[PublicKey] peers, ref const QuorumConfig config)
```

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -159,6 +157,42 @@ extern(D):
+                getNodes(sub_conf, nodes);
+        }
+
+        // can't use Set(), requires serialization support
```
That not true anymore is it ?

Geod24 replied:
Especially since `Set` is imported

AndrejMitrovic commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -159,6 +157,42 @@ extern(D):
+                getNodes(sub_conf, nodes);
+        }
+
+        // can't use Set(), requires serialization support
```
Aha, I've just noticed this is no longer true! Yay refactorings

Geod24 replied:
And yay self reviews!

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -277,44 +277,54 @@ public class Ledger
 
+    ***************************************************************************/
+
+    private bool prepareNominatingSet (ref Set!Transaction txs) @safe
```
Is there ever a chance we nominate an empty set ? If not we can just simplify this signature...

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -277,44 +277,54 @@ public class Ledger
+        // note: we are not passing the previous tx set as we don't really
+        // need it at this point (might later be necessary for chain upgrades)
+        auto slot_idx = this.last_block.header.height + 1;
+        this.nominate(slot_idx, Set!Transaction.init, txs);
```
Looks like we might want to change the signature of `NominateDg` at some point if it's always called with `.init` as second arg ?

AndrejMitrovic replied:
This delegate is removed in a further commit. I originally had it part of one commit, but the changes were very hard to review, so I split it up.

Geod24 reviewed (commented):
First round of review done, will need another one for the last two commits

AndrejMitrovic committed:
setupNetwork: Use Set!() in place of hashmap

AndrejMitrovic commented:
- Rebased against master
- Fixed definition of `prepareNominatingSet` so it doesn't needlessly return a boolean status flag.
- Replaced hashes with sets.
- Added const.

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -91,17 +95,18 @@ extern(D):
         this.key_pair = key_pair;
         auto node_id = NodeID(StellarHash(key_pair.address[]));
         const IsValidator = true;
+        auto quorum_set = verifyBuildSCPConfig(config);
```
Pass it directly on the next line to avoid a copy.

AndrejMitrovic replied:
I can't, it takes a reference.

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -158,6 +156,42 @@ extern(D):
+
+    public void setupNetwork (NetworkClient[PublicKey] clients)
+    {
+        import agora.common.Set;
```
This import is redundant

Geod24 reviewed (approved):
LGTM, only 2 nits. Feel free to self-merge after addressing them (and rebasing, sorry :D )
